### ISSUE TITLE
When you invite a user who is not registered, it should register him automatically when they click the link

### ISSUE THREAD

**Issue by pietrosperoni** [fairdemocracy/vilfredo-client#125]
So, I create a first user: Pietro, email nightly@...
Pietro receives the email and signs in
Pietro creates a question: question1
Pietro invites his friend nightly2@...
On another browser I receive the following email:
![screen shot 2015-01-17 at 10 33 39](https://cloud.githubusercontent.com/assets/1333674/5788538/b16cb646-9e33-11e4-8699-23109989ad0c.png)

> If you are already a member of Vilfredo please login then click on the link below. 

... (no problem here)

> If you are not yet registered then please go to http://nightly.vilfredo.org and register for a free account. Once you have logged in, click on the link below.

This is problematic because most people do not read the text of emails, they simply click on the link. And I did so as well, which lead me to nightly.vilfredo.org where I discovered I was not registered so I registered. And then it sent me a second email to nightly2 (even though the server already knew that email was mine). Once I was registered, verified (uselessly), logged in (also uselessly), I now could not see the question because the link from the first invitation email has been burned.

Worse than that when Pietro invited nightly2 again, the answer was: 
![screen shot 2015-01-17 at 10 39 43](https://cloud.githubusercontent.com/assets/1333674/5788552/8ea9515e-9e34-11e4-9902-cd4b3e28acb9.png)
 
Notice that instead is I try to invite someone exnovo the message is Emails sent:
![screen shot 2015-01-17 at 10 40 35](https://cloud.githubusercontent.com/assets/1333674/5788556/c8850a94-9e34-11e4-800a-3c284c1c7f67.png)

So, inviting a person to a question who is not registered does not work because people will simply ignore the test of the email, and this will make it impossible for them to be invited. Unfortunately this is not something we can postpone to the next version. So although pressed this needs to be done before the test.

SOLUTION:

If you invite someone who is not registered, the mail should have a single click (as it has). 
If you are registered it will add you to the question
If you are NOT registered it will bring you to a page where you can write your username, the email is already written, ask your password, and then you click to register. Next to the email is says that if you want you are allowed to change your email *after* . Once you click register, you do NOT receive an email (we already know you control that email). You are directly added to that question. And you are directly logged in. You do not need to do anything else.

@giovanisp any comment?
@vvorski you are a great fun of zero friction UI. Would this look ok.

Please all consider that the milestone is due in very few days. So this is urgent work.
( @giovanisp , what do you think, if this is not done, should we postpone the release? We really cannot go out with this not working)

**Comment by pietrosperoni** [comment]
Notice that when I re-invite a person, Vilfredo correctly notices that I have already invited him:
![screen shot 2015-01-17 at 10 50 27](https://cloud.githubusercontent.com/assets/1333674/5788571/20678db2-9e36-11e4-802a-a2617f54c97b.png)

But it does not invite him again. And yet in the participant that person is not there, and there is no way for him to accept that invitation once he has burned the link.

![screen shot 2015-01-17 at 10 52 13](https://cloud.githubusercontent.com/assets/1333674/5788573/491d8a4a-9e36-11e4-87c3-8c766ff5bc5f.png)

**Comment by giovanisp** [comment]
Yes this is high prio. Frictionless invitation system.

The solutions is great. The ideal one should have the multiple email management. But yours looks pretty straightforward: 

"If you invite someone who is not registered, the mail should have a single click (as it has). 
If you are registered it will add you to the question

If you are NOT registered it will bring you to a page where you can write your username, the email is already written, ask your password, and then you click to register. Next to the email is says that if you want you are allowed to change your email after . 

Once you click register, you do NOT receive an email (we already know you control that email). GS NOTE: we can still send him an email so the person knows his account is activated.

You are directly added to that question. And you are directly logged in. You do not need to do anything else."

Gs


> Il giorno 17/gen/2015, alle ore 11:44, Pietro Speroni <notifications@github.com> ha scritto:
> 
> If you invite someone who is not registered, the mail should have a single click (as it has). 
> If you are registered it will add you to the question
> If you are NOT registered it will bring you to a page where you can write your username, the email is already written, ask your password, and then you click to register. Next to the email is says that if you want you are allowed to change your email after . Once you click register, you do NOT receive an email (we already know you control that email). You are directly added to that question. And you are directly logged in. You do not need to do anything else.

**Comment by derekpa** [comment]
Of course once we have a notification system the newly registered user would also get the notification of the invite.

However, it seems to me the simplest and quickest way to fix this is :

1) if someone then registers with that same email they were invited with the system can easily 

a) automatically accept the invite(s) associated with the email address 

or 

b )send a second email along the lines of: OK, now you've registered please click this link if you want to accept this invitation.

**Comment by derekpa** [comment]
@pietrosperoni OK what you say here makes no sense.

"And I did so as well, which lead me to nightly.vilfredo.org where I discovered I was not registered so I registered. And then it sent me a second email to nightly2 (even though the server already knew that email was mine). Once I was registered, verified (uselessly), logged in (also uselessly), I now could not see the question because the link from the first invitation email has been burned."

Why "uselessly"? Just because there is an invite to nightly2@..... does not mean the system should take your word that you control the address when you signed up!!!

Or am I missing something here?

Also, clicking on the link after activating your account should have worked - the flow you outlined worked before many times. I shall test it again.

**Comment by pietrosperoni** [comment]
When a user is not register, and receives an email, he should have a link. And clicking on the link the system knows what invitation is this coming from, and knows that no user is associated to that email. So Vilfredo knows that whoever clicked that link had access to that email. So there is no need to test it again.

It just needs to get a username, a password. And he is good to go. The new user also will be already included in the question.

So if a person if not registered:
he receives an email, goes to the page, the system knows he is coming from that invitation, registers him, and that is all.

I will test the link that did not work. But independently of that, people should not need to go through a formal registration if they come there following a link in an email.

**Comment by derekpa** [comment]
Yes, I realised after I wrote it what you meant. It does make sense the way it's done, but there's no need if we make just one link.

**Comment by pietrosperoni** [comment]
Yes, you are right. The link works. But still is unnecessarily long.

**Comment by derekpa** [comment]
Yes, it's just the way I added it, for quickness. But it should easily be just click one link.

**Comment by pietrosperoni** [comment]
Now is:
user 1 (u1) sends an inviations
user 2 (u2) receives the invitation
u2 clicks on the link
u2 registers putting username, email, and password.
u2 goes to the email and proves that it's his
u2 logs in (maybe this is needed right now maybe not. Surely eventually we can log them in automatically)
u2 opens the old email again
u2 clicks on the link and signs in to the question
u1 receives an email that u2 joined

Instead it should be:
user 1 (u1) sends an inviations
user 2 (u2) receives the invitation
u2 clicks on the link
u2 registers putting username, and password.
u1 receives an email that u2 joined


c'est plus facil ;-)

**Comment by pietrosperoni** [comment]
In fact it should be:
user 1 (u1) sends an inviations
user 2 (u2) receives the invitation
u2 clicks on the link
U2 says he is not registered yet and registers putting username, and password.
u1 receives an email that u2 joined

while if u2 was registered with another mail it should:
user 1 (u1) sends an inviations
user 2 (u2) receives the invitation
u2 clicks on the link
U2 says he is already registered and logs in. Vilfredo assigns that email as another email of that person.
u1 receives an email that u2 joined

**Comment by giovanisp** [comment]
Yes. Let's just make sure if the "multiple email scenario" can be done now or has to be postponed (there's another issue for this).

**Comment by derekpa** [comment]
Well do we really want to store multiple emails for a user? I wouldn't want that. I like that I can receive an email to any address to join a question, but only have one address associated with/ stored in Vilfredo.

**Comment by pietrosperoni** [comment]
People use multiple email. And they can be invited using multiple emails. So yes. one person, many emails. Like on facebook, and linkedin and a lot of the other serious sites.

**Comment by pietrosperoni** [comment]
@giovanisp I understand. But it is really important that people can log in easily when receiving an invitation. 
How mong would it take you, @derekpa , to program it?

**Comment by pietrosperoni** [comment]
And yes, indeed you should define which one is your primary email

**Comment by derekpa** [comment]
No long to add it. Do it as part of a day's work. 

But I imagine we should be asking whether the user wants the system to save all these email addresses....? Not everyone is to deep into social media that thy're happy to have websites accumulate personal data. It's a privacy thing above all else, and many users will undoubtedly not want this.

**Comment by pietrosperoni** [comment]
We can let them delete it later. People say they want privacy but every single time they have to chose between convenience and privacy the great majority chooses convenience. So we can let people delete an email address. But the default must address what the majority of people want.

@giovanisp , what's your take?

**Comment by giovanisp** [comment]
See #127 . I agree on linking multiple email addresses and setting a primary email (linkedIn style). Not sure if it is doable in 2 days. We need to wrap-up 2.0.0-rc1.

**Comment by pietrosperoni** [comment]
Yes, I am also giving a presentation on Saturday for a community here in Berlin

**Comment by pietrosperoni** [comment]
YES! This works! Great. 
(I had uncovered another bug in the process: #134  , but this one works. To test it, I had to make the font size tiny to work around the bug above).
